src/libopensc/card-piv.c

```diff
@@ -2992,19 +3032,47 @@ static int piv_match_card(sc_card_t *card)
 	if (card->type == -1)
 		card->type = type;
 
-	card->drv_data = priv;
+	card->drv_data = priv; /* will frre if no match, or pass on to piv_init */
 	priv->aid_file = sc_file_new();
 	priv->selected_obj = -1;
 	priv->pin_preference = 0x80; /* 800-73-3 part 1, table 3 */
 	priv->logged_in = SC_PIN_STATE_UNKNOWN;
 	priv->tries_left = 10; /* will assume OK at start */
+	priv->pstate = PIV_STATE_MATCH;
 
 	/* Some objects will only be present if Histroy object says so */
 	for (i=0; i < PIV_OBJ_LAST_ENUM -1; i++)
 		if(piv_objects[i].flags & PIV_OBJECT_NOT_PRESENT)
 			priv->obj_cache[i].flags |= PIV_OBJ_CACHE_NOT_PRESENT;
 
+	sc_lock(card);
 
+	/*
+	 * detect if active AID is PIV. NIST 800-73 says Only one PIV application per card
+	 * and PIV must be the default application
+	 * This can avoid doing doing a select_aid and losing the login state on some cards
+	 * We may get interference on some cards by other drivers trying SELECT_AID before
+	 * we get to see if PIV application is still active.
+	 * putting PIV driver first might help. 
+	 * TODO could be cached too
+	 */
+	 i = piv_find_discovery(card);
+
+	if (i < 0) {
+	    /* Detect by selecting applet */
+	    i = piv_find_aid(card, &aidfile);
+	}
+
+	if (i < 0) {
+		piv_finish(card);
+		 /* don't match. Does not have a PIV applet. */
+		 sc_unlock(card);
+		 return 0;
```

Comment:
bad indenting

Reply:
OK, fixed and will be in next commit. 
Note I did a commit earlier to handle the CKU_CONTEXT_SPECIFIC login.

Reply:
Still I see there one space after the two tabs in the above three lines. The block above is indented with tab and 4 spaces.